README.md
[![manaakiwhenua-standards](https://github.com/manaakiwhenua/compGeometeR/workflows/manaakiwhenua-standards/badge.svg)](https://github.com/manaakiwhenua/manaakiwhenua-standards)



# compGeometeR

CompGeometeR aims to expose the implementation of various computation geometry algorithms in R. We believed by exposing these algorithms, the R user could easily apply these algorithms in solving problems in different domains. For a start we have exposed some computation geometry implemented in the [Qhull library](http://www.qhull.org/) and implement additional algorithms using the output derived from the Qhull implemented algorithms. The following algorithm are available for use:

* Delunay: Computes the [Delaunay triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation) by computing a [convex hull](https://en.wikipedia.org/wiki/Convex_hull)

* Convex Hull : Implements the Quickhull algorithm for computing a [convex hull](https://en.wikipedia.org/wiki/Convex_hull). This is done using the [Qhull library](http://www.qhull.org/)

* In-Convex: Given a point with n-dimensional, the goal is to check which of the set of a test point are within the [convex hull](https://en.wikipedia.org/wiki/Convex_hull}{convex hull) 

* Alpha-Complex: Calculates the [alpha complex](href{https://en.wikipedia.org/wiki/Alpha_shape#Alpha_complex) of a set of point in n-dimensional using the [Qhull library](http://www.qhull.org/)

* Find simplex: Compute the simplices of a [Delaunay triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation) or [alpha complex](href{https://en.wikipedia.org/wiki/Alpha_shape#Alpha_complex) that contain the given set of test points.


* Voronoi: Computes the [voroni diagram](https://en.wikipedia.org/wiki/Voronoi_diagram) for a given set of point using [Qhull library](http://www.qhull.org/)




## Learning More

To learn more about computational geometry library/algorithms and R package development, visit the following:


*  To learn about R development, please see the chapter[Whole Game](http://r-pkgs.org/whole-game.html) in the R Packages development book. It gives a comprehensive explanation for R package development and uses the devtools package. Also, [chapeter 4](http://r-pkgs.org/package-structure-state.html) of that book gives a great explanation of how to structure your R package.

* [Qhull library](http://www.qhull.org/) for computational algorithm used


* Convex Hull: references Barber CB, Dobkin DP, Huhdanpaa H (1996) The Quickhull algorithm for convex hulls. ACM Transactions on Mathematical Software, 22(4):469-83  url:https://doi.org/10.1145/235815.235821.


* Alpha Shape: Edelsbrunner H, Mücke EP (1994) Three-dimensional alpha shapes. ACM Transactions on Graphics, 13(1):43-72             url:https://dl.acm.org/doi/abs/10.1145/174462.156635. 






## Citation



## Community Guidelines
For bugs or request to add extra functionality please use the [issue](https://github.com/manaakiwhenua/compGeometeR/issues) functionality to get in touch 


## Dependencies

`compGeometeR` dependants on `devtools` package and `R` version 3.5.1. was used for development

## Quick Examples in Ecology Domain



``` r
library(compGeometeR)


```

<img src="man/figures/README-1.png" width="100%" />




## Github Installation

```r
# Install devtools from CRAN
install.packages("devtools")
devtools::install_git("https://github.com/manaakiwhenua/compGeometeR")

``` 
    

## Local Development

This software was developed using `R` and `C` Language. If you fork this project and want to extend the current functionality, you can do that by extending the`R`code or `C` code or both. To get started you need the following software:

### Software
1. Download [perl](https://www.activestate.com/products/activeperl/downloads/)


2. Download make using the step below


  * Download make.exe from their [official site]("http://gnuwin32.sourceforge.net/packages/make.htm")
  
  * In the download session, click complete package, except sources.
  
  * Follow the installation instructions.
  
  * Once finished, add the <installation directory>/bin/ to the PATH variable.



3. Download [html tidy](http://www.paehl.com/open_source/?HTML_Tidy_for_Windows). for generating a tidy help package


4. Download R studio and use it to open the .Rproj


5. Use Eclipse or any C editor to update the C files as needed.


### Building and Installing

There are two ways to build and install the package. This is done by using the make file or by using the `devtools` package in R. 
The `make` file is the easier way to rebuild and install the package, this is because it creates the documentation, builds, and installs the package using one command. To do this, please run the following command in the package root directory: 


 * make install (this should be run outside the project directory)

To build and install using the R devtools, you have to make sure you `make` install on your system. Once you have follow the instruction given above to install `make`, then run the following.

 * roxygen2::roxygenise()  (to regenrate the documentation)
 * devtools::build()  (to build)
 * devtools::install or devtools::install(args = c("--no-multiarch"))  (to install depending on your architecture)

### Generating Package PDF

To re-generate the package pdf, please run the following
 * R CMD Rd2pdf --title='compGeometeR' -o /path-to-your-project-folder/man/*.Rd 





## Maintainer
Pascal Omondiagbe <omondiagbep@landcareresearch.co.nz>






NEWS.md
# compGeomterR 1.0
, 'alpha_complex'
1. First release. `in-convex-hull` ,`convex-hull`,'convex_layer', `delaunay`,`find_simplex`,`grid_coordinates`,`voronoi` and `alpha-complex`


R/alpha-complex.R
#' @title Alpha complex
#' 
#' @description  This function calculates the 
#' \href{https://en.wikipedia.org/wiki/Alpha_shape#Alpha_complex}{alpha complex} 
#' of a set of \eqn{n} points in \eqn{d}-dimensional space using the
#' \href{http://www.qhull.org}{Qhull} library.
#' 
#' @param points a \eqn{n}-by-\eqn{d} dataframe or matrix. The rows
#'   represent \eqn{n} points and the \eqn{d} columns the coordinates in 
#'   \eqn{d}-dimensional space.
#' @param alpha a real number between zero and infinity that defines the maximum 
#'   circumradii for a simplex to be included in the alpha complex.  If 
#'   unspecified \code{alpha} defaults to infinity and the alpha complex is 
#'   equivalent to a Delaunay triangulation.
#' 
#' @return Returns a list consisting of:
#' 
#' \itemize{
#'   \item \code{input_points}: the input points used to create the Voronoi 
#'   diagram.
#'   \item \code{simplices}: a \eqn{s}-by-\eqn{d+1} matrix of point indices 
#'   that define the \eqn{s} \href{https://en.wikipedia.org/wiki/Simplex}{simplices} 
#'   that make up the alpha complex.
#'   \item \code{circumcentres}: a \eqn{s}-by-\eqn{d} matrix of coordinates 
#'   that define the centre of the 
#'   \href{https://en.wikipedia.org/wiki/Circumscribed_circle}{circumcircle} 
#'   associated with each simplex.
#'   \item \code{circumradii}: the radius of each circumcircle.
#' }
#' 
#' @references Barber CB, Dobkin DP, Huhdanpaa H (1996) The Quickhull algorithm 
#' for convex hulls. ACM Transactions on Mathematical Software, 22(4):469-83 
#' \url{https://doi.org/10.1145/235815.235821}.
#' 
#' Edelsbrunner H, Mücke EP (1994) Three-dimensional alpha shapes. ACM 
#' Transactions on Graphics, 13(1):43-72 
#' \url{https://dl.acm.org/doi/abs/10.1145/174462.156635}.
#' 
#' @examples 
#' # Define points
#' x <- c(30, 70, 20, 50, 40, 70)
#' y <- c(35, 80, 70, 50, 60, 20)
#' p <- data.frame(x, y)
#' # Create alpha complex and plot
#' a_complex <- alpha_complex(points = p, alpha = 20)
#' plot(p, pch = as.character(seq(nrow(p))), xlim=c(0,80), ylim=c(10,90), asp=1)
#' for (s in seq(nrow(a_complex$simplices))) {
#'   polygon(a_complex$input_points[a_complex$simplices[s,],], border="red")
#' }
#' text(a_complex$circumcentres, labels=seq(nrow(a_complex$simplices)), col="blue")
#' symbols(a_complex$circumcentres, circles = a_complex$circumradii, 
#'         inches = FALSE, add = TRUE, fg="blue", lty="dotted")
#' 
#' @export
alpha_complex <- function(points=NULL, alpha=Inf) {
	
    # Check directory writable
    tmpdir <- tempdir()
    # R should guarantee the tmpdir is writable, but check in any case
    if (file.access(tmpdir, 2) == -1) {
      stop(paste("Unable to write to R temporary directory", tmpdir, "\n"))
    }
        
    # Coerce the input to be matrix
    if(is.null(points)){
      stop(paste("points must be an n-by-d dataframe or matrix", "\n"))
    }
    if(!is.data.frame(points) & !is.matrix(points)){
      stop(paste("points must be a dataframe or matrix", "\n"))
    }
    if (is.data.frame(points)) {
      points <- as.matrix(points)
    }
    # Make sure we have real-valued input
    storage.mode(points) <- "double"
    # We need to check for NAs in the input, as these will crash the C code.
    if (any(is.na(points))) {
      stop("points should not contain any NAs")
    }
    
    # Specify the Qhull options: http://www.qhull.org/html/qh-optq.htm
    if (ncol(points) < 4) {
      options <- "Qt Qc Qz"
    } else {
      options <- "Qt Qc Qx"
    }
    options <- paste(options, collapse=" ")  

	  # Call C function to create the Voronoi diagram
  	vd <- .Call("C_voronoiR", points, options, tmpdir, PACKAGE="compGeometeR")
    # Re-index from C numbering to R numbering
    vd$tri[is.na(vd$tri)] <- 0
    tri <- vd$tri + 1
    
    # Create list to return the desired alpha complex information
    alpha_complex <- list()
    alpha_complex$input_points <- points
    in_alpha_complex <- vd$circumRadii <= alpha
    alpha_complex$simplices <- tri[in_alpha_complex, ]
    
    if (nrow(alpha_complex$simplices) < 1) {
	    alpha_complex$circumcentres <- NULL
	    alpha_complex$circumradii <- NULL
	    } else {
	    alpha_complex$circumcentres <- vd$voronoi_vertices[in_alpha_complex,]
	    alpha_complex$circumradii <- vd$circumRadii[in_alpha_complex]
	  }

  	return(alpha_complex)
  }
